pkg/azure/api/api.go

```diff
@@ -148,96 +181,136 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context) ([]network.Inter
+	for _, virtualMachineScaleSet := range virtualMachineScaleSets {
+		virtualMachineScaleSetNetworkInterfaces, err := c.listVirtualMachineScaleSetNetworkInterfaces(ctx, *virtualMachineScaleSet.Name)
 		if err != nil {
 			return nil, err
 		}
 
-		scaleset := result.Value()
-		err = result.Next()
+		networkInterfaces = append(networkInterfaces, virtualMachineScaleSetNetworkInterfaces...)
+	}
 
-		if scaleset.Name == nil {
-			continue
+	return networkInterfaces, nil
+}
+
+// listVirtualMachineScaleSets lists all virtual machine scale sets in the client's resource group
+func (c *Client) listVirtualMachineScaleSets(ctx context.Context) ([]armcompute.VirtualMachineScaleSet, error) {
+	var virtualMachineScaleSets []armcompute.VirtualMachineScaleSet
+	var err error
+
+	c.limiter.Limit(ctx, virtualMachineScaleSetsList)
+	sinceStart := spanstat.Start()
+
+	pager := c.virtualMachineScaleSets.NewListPager(c.resourceGroup, nil)
+
+	defer func() {
+		c.metricsAPI.ObserveAPICall(virtualMachineScaleSetsList, deriveStatus(err), sinceStart.Seconds())
+	}()
+
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, err
+		}
+		for _, virtualMachineScaleSet := range nextResult.Value {
+			if virtualMachineScaleSet.Name == nil {
+				continue
+			}
+			virtualMachineScaleSets = append(virtualMachineScaleSets, *virtualMachineScaleSet)
 		}
+	}
+
+	return virtualMachineScaleSets, nil
+}
+
+// listVirtualMachineScaleSetNetworkInterfaces lists all network interfaces for a given virtual machines scale set
+func (c *Client) listVirtualMachineScaleSetNetworkInterfaces(ctx context.Context, virtualMachineScaleSetName string) ([]armnetwork.Interface, error) {
+	var networkInterfaces []armnetwork.Interface
+	var err error
+
+	c.limiter.Limit(ctx, interfacesListVirtualMachineScaleSetNetworkInterfaces)
+	sinceStart := spanstat.Start()
 
-		c.limiter.Limit(ctx, InterfacesListAll)
-		sinceStart := spanstat.Start()
-		result2, err2 := c.interfaces.ListVirtualMachineScaleSetNetworkInterfacesComplete(ctx, c.resourceGroup, *scaleset.Name)
-		c.metricsAPI.ObserveAPICall(InterfacesListVirtualMachineScaleSetNetworkInterfacesComplete, deriveStatus(err2), sinceStart.Seconds())
-		if err2 != nil {
+	pager := c.interfaces.NewListVirtualMachineScaleSetNetworkInterfacesPager(c.resourceGroup, virtualMachineScaleSetName, nil)
+
+	defer func() {
+		c.metricsAPI.ObserveAPICall(interfacesListVirtualMachineScaleSetNetworkInterfaces, deriveStatus(err), sinceStart.Seconds())
+	}()
+
+	for pager.More() {
+		nextResult, err := pager.NextPage(ctx)
+
+		if err != nil {
 			// For scale set created by AKS node group (otherwise it will return an empty list) without any instances API will return not found. Then it can be skipped.
-			var v autorest.DetailedError
-			if errors.As(err2, &v) && v.StatusCode == http.StatusNotFound {
+			var respErr *azcore.ResponseError
+			if errors.As(err, &respErr) && respErr.StatusCode == http.StatusNotFound {
```

Comment:
Bit confused by this snippet (before and after the change). Won't `respErr` just always be nil here, causing errors.As(err, &respErr) to just always be false?

Reply:
I don't think so, this looks consistent with [this example snippet](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore#example-ResponseError): when `err` is not nil, we use [errors.As](https://pkg.go.dev/errors#As) to cast it to `respErr` which is a `*azcore.ResponseError`. If that cast is successful, then `errors.As(...)` returns `true` and `respErr` is no longer a nil pointer.
> As finds the first error in err's tree that matches target, and if one is found, sets target to that error value and returns true. Otherwise, it returns false.

As for `errors.As(...) && respErr.xxx`, errors.As is evaluated first, so by the time `respErr.StatusCode` is evaluated, `respErr` is no longer nil (see [this example snippet](https://go.dev/play/p/d0zeybzVGcq))